build/utf8/valid_asm.go

```diff
@@ -0,0 +1,447 @@
+		TooLong | Overlong2 | TwoConts | Surrogate | TooLarge,
+
+		// ________ 11______
+		TooShort, TooShort, TooShort, TooShort,
+	})
+
+	return
+}
+
+func main() {
+	TEXT("validateAvx", NOSPLIT, "func(p []byte) (bool, bool)")
+	Doc("Optimized version of Validate for inputs of more than 32B.")
+
+	retUtf8, _ := ReturnIndex(0).Resolve()
+	retAscii, _ := ReturnIndex(1).Resolve()
+
+	d := Load(Param("p").Base(), GP64())
+	n := Load(Param("p").Len(), GP64())
+
+	validAsciiReg := GP8()
+	MOVB(Imm(1), validAsciiReg)
+
+	JumpIfFeature("init_avx", cpu.AVX2)
+
+	// TODO: call stdlib
+
+	Label("init_avx")
+
+	Comment("Prepare the constant masks")
+
+	incompleteMask := ConstBytes("incomplete_mask", incompleteMaskData())
+	incompleteMaskY := YMM()
+	VMOVDQU(incompleteMask, incompleteMaskY)
+
+	continuation4Bytes := ConstBytes("cont4_vec", continuationMaskData(0b11110000))
+	continuation4BytesY := YMM()
+	VMOVDQU(continuation4Bytes, continuation4BytesY)
+
+	continuation3Bytes := ConstBytes("cont3_vec", continuationMaskData(0b11100000))
+	continuation3BytesY := YMM()
+	VMOVDQU(continuation3Bytes, continuation3BytesY)
+
+	nib1Data, nib2Data, nib3Data := nibbleMasksData()
+
+	Comment("High nibble of current byte")
+	nibble1Errors := ConstBytes("nibble1_errors", nib1Data)
+	nibble1Y := YMM()
+	VMOVDQU(nibble1Errors, nibble1Y)
+
+	Comment("Low nibble of current byte")
+	nibble2Errors := ConstBytes("nibble2_errors", nib2Data)
+	nibble2Y := YMM()
+	VMOVDQU(nibble2Errors, nibble2Y)
+
+	Comment("High nibble of the next byte")
+	nibble3Errors := ConstBytes("nibble3_errors", nib3Data)
+	nibble3Y := YMM()
+	VMOVDQU(nibble3Errors, nibble3Y)
+
+	Comment("Nibble mask")
+	lowerNibbleMask := ConstArray64("nibble_mask",
+		0x0F0F0F0F0F0F0F0F,
+		0x0F0F0F0F0F0F0F0F,
+		0x0F0F0F0F0F0F0F0F,
+		0x0F0F0F0F0F0F0F0F,
+	)
+
+	nibbleMaskY := YMM()
+	VMOVDQU(lowerNibbleMask, nibbleMaskY)
+
+	Comment("For the first pass, set the previous block as zero.")
+	previousBlockY := YMM()
+	zeroOutVector(previousBlockY)
+
+	Comment("Zeroes the error vector.")
+	errorY := YMM()
+	zeroOutVector(errorY)
+
+	Comment(`Zeroes the "previous block was incomplete" vector.`)
+	incompletePreviousBlockY := YMM()
+	zeroOutVector(incompletePreviousBlockY)
+
+	Comment("Top of the loop.")
+	Label("check_input")
+
+	Comment("if bytes left >= 32")
+	CMPQ(n, U8(32))
+	Comment("go process the next block")
+	JGE(LabelRef("process"))
+
+	Comment("If < 32 bytes left")
+
+	Comment("Fast exit if done")
+	CMPQ(n, U8(0))
+	JE(LabelRef("end"))
+
+	Comment("If 0 < bytes left < 32.")
+
+	Comment("Prepare scratch space")
+	scratch := AllocLocal(32)
+	scratchAddr := GP64()
+	LEAQ(scratch, scratchAddr)
+	tmpZeroY := zeroOutVector(YMM())
+	VMOVDQU(tmpZeroY, Mem{Base: scratchAddr})
+
+	Comment("Make a copy of the remaining bytes into the zeroed scratch space and make it the next block to read.")
+	copySrcAddr := GP64()
+	MOVQ(d, copySrcAddr)
+	MOVQ(scratchAddr, d)
+
+	copyN(d, copySrcAddr, n)
+	MOVQ(U64(32), n)
```

Comment:
If you use blend, you'll also have to handle the case where the input crosses a page boundary. See https://github.com/segmentio/asm/blob/248f170470e61e18aa8ebd6eaf82e1714fff1c21/build/keyset/keyset_asm.go#L55-L59 and https://github.com/segmentio/asm/blob/248f170470e61e18aa8ebd6eaf82e1714fff1c21/build/keyset/keyset_asm.go#L149-L177 for an example, and https://github.com/segmentio/asm/blob/248f170470e61e18aa8ebd6eaf82e1714fff1c21/keyset/keyset_test.go#L65-L96 for an example for testing loads around a page boundary.

Reply:
Thanks! I'm trying to understand how this would work for a load of 16 <= N < 32 bytes. Given the trick uses a mask of twice the max size of key, are you expecting to do two 16 bytes loads and apply this method the blend to the second one?

Reply:
The keyset version uses 16 byte XMM registers because that's all that's needed. You should be able to do a 32 byte load into a YMM register, and then use the [AVX2 version of PBLENDVB](https://www.felixcloutier.com/x86/pblendvb) which lets you blend 32 bytes at a time. Using the trick above, you'd load the 32 byte blend mask from a 64 byte region.

Reply:
How would it work for a load close to a page boundary though?

For example, let's say there are 19 bytes of input remaining, ending at a page boundary.

I expect the result of a 32 bytes load up to and including the remainder of the input would look like:

```
0                          104   128                              256
[.12.bytes.of.garbage......rest..|.of.input.......................]
                           ^
```

Where `garbage` really is the previous block of data. Using `VPBLENDVB` I can shift right the first half of the vector by 12 bytes, but how can I also shift the second half? It would need to cross lanes, but we can't use the `VPERM2I128` + `VPALIGNR` directly, as the latter takes an immediate only as offset. I guess we could generate all 15 variations of VPALIGNR and jump to the the right one, but it seems like I'm missing something :/

Reply:
Ah yeah, the cross page boundary load is a different story. I was talking about blending in the zeroes once you've aligned the input. For load+shuffle, PSHUFB doesn't let you shuffle across lanes, so you'd need a combination of PERM2I128+PSHUFB+PLENDVB to get the desired result.

Reply:
If you had 18 bytes of input like:

```
| .... .... .... XXXX | YYYY YYYY YYYY ZZZZ |
```

and you wanted to shift it and pad with zeroes:

```
| XXXX YYYY YYYY YYYY | ZZZZ 0000 0000 0000 |
```

You could do something like:
- build `tmp1` via VPSHUFB (or VPALIGNR?) => `| XXXX .... .... .... .... | .... .... .... .... |`
- build `tmp2` via VPERM2I128 => `| YYYY YYYY YYYY ZZZZ | .... .... .... .... |`
- build `tmp2b` via VPSHUFB => `| .... YYYY YYYY YYYY | .... .... .... .... |`
- build `tmp3` via VPSHUFB (or VPALIGNR?) => `| .... .... .... .... | ZZZZ .... .... .... |`
- blend tmp1+tmp2b+tmp3+zeroes to obtain the final result

Reply:
Oh right, thank you! Got something working that way. Also `tmp1` and `tmp2` are really the same vector.

Reply:
https://github.com/segmentio/asm/pull/58/commits/aa912a4e7447f8c5b7408956530e6108b6ff466a